### PR TITLE
feat: rewrite syschdemd

### PR DESCRIPTION
**Pull request by K900**
Mostly just a big pile of cleanups.

Important changes:
- refactor and clean up the code, fix all the shellcheck complaints
- stop trying to guess paths in scripts and wrap them correctly instead
- use nsfs instead of trying to figure out the right PID to copy namespaces from
- clean up $WSLPATH to remove extra impure entries
- don't try to restart systemd if it died
- make sure the store is read-only before we do anything, so systemd can't mess with it

### CODE REVIEW

File: scripts/syschdemd.nix [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+, lib
+, coreutils
+, daemonize
+, glibc
+, gnugrep
+, systemd
+, util-linux
+, defaultUser
+, automountPath
+,
```
Comment (SuperSandro2000) [inline scripts/syschdemd.nix:12]:
```suggestion
```

File: scripts/syschdemd.nix [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+}:
+let
+  writeShellScript =
```
Comment (SuperSandro2000) [inline scripts/syschdemd.nix:15]: This shadows a function in nixpkgs https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/trivial-builders.nix#L213

File: scripts/syschdemd.sh [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        machinectl \
+            --quiet \
+            --uid=@defaultUser@ \
+            --setenv=INSIDE_NAMESPACE=true \
+            --setenv=WSLPATH="$(clean_wslpath)" \
+            shell .host \
+                /bin/sh -c "cd \"$PWD\"; $exportCmd; source /etc/set-environment; exec $command"
+}
+
+main "$@"
```
Comment (SuperSandro2000) [inline scripts/syschdemd.sh:134]: Missing final new line

File: scripts/syschdemd.sh [hunk 4]
```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+[ "${NIXOS_WSL_DEBUG:-}" == "1" ] && set -x
```
Comment (SuperSandro2000) [inline scripts/syschdemd.sh:4]:
```suggestion
[[ -v NIXOS_WSL_DEBUG ]] && set -x
```
testing if it is set to anything maybe makes it easier if people accidentally use true

File: scripts/syschdemd.nix (hunk 1 above)
Comment (K900) [reply scripts/syschdemd.nix:12]: Fixed.

File: scripts/syschdemd.sh (hunk 3 above)
Comment (K900) [reply scripts/syschdemd.sh]: I just reformatted the whole thing with shfmt.

Comment (SuperSandro2000) [reply scripts/syschdemd.sh]: Just for my interest: What settings are you using? Maybe we should document that somewhere to not reformat files multiple times. I am always using ``shfmt -ci -i 2 -s -w $FILE``.

Comment (K900) [reply scripts/syschdemd.sh]: I did add it as a check in CI.

**Comment by nzbr** [comment]
Docker Desktop is broken, because the `mount --make-rshared ${automountPath}` line is missing from the systemd wrapper script. It's needed so that the bind mount that docker creates is present outside of the systemd container

**Comment by nzbr** [comment]
Other than that, everything seems to be working

**Comment by K900** [comment]
I tried adding it to the _outside_ script, can you see if that works?

**Comment by K900** [comment]
I am very confused by how the mounts work then. Can you run `findmnt -o+PROPAGATION` outside the container?

**Comment by K900** [comment]
Is this from inside the namespace or outside?

**Comment by nzbr** [comment]
The command output is from inside the systemd namespace, but not the docker container. Outside the systemd namespace it looks like this:

[mounts-outside.txt](https://github.com/nix-community/NixOS-WSL/files/9426267/mounts-outside.txt)

**Comment by nzbr** [comment]
This is also a problem for other mounts created inside the namespace: I have a tmpfs on /tmp in my NixOS installation (with the current syschdemd). Because of that any files in /tmp can't be accessed from outside it (e.g. through SMB with the explorer or interop) and trying to use interop in there results in a not at all helpful error:
image

When I made the docker-desktop fix, I also tried removing mount namespacing completely, but that means that the namespace and the ouside have the same `/proc` filesystem which breaks even more things

**Comment by K900** [comment]
Can you try the latest commit? I think unshare was just doing a weird.

**Comment by nzbr** [comment]
The latest commit seems to be broken
image

**Comment by K900** [comment]
Forgot to remove the other thing. Can you try it now?

**Comment by nzbr** [comment]
image

Seems to be working now

**Comment by K900** [comment]
Squashed, should be good to go now.

**Pull request merged**